### PR TITLE
[AMD] Add Optimization for DeepSeek-R1/V3 based on AITER Backend

### PR DESCRIPTION
**Pull request by BruceXcluding**
## Motivation
To optimize the high-performance solution for DeepSeek-R1/V3, particularly focusing on aiter backend. Below is a structured optimization path. wait for cleaning up the aiter branch.

## Modifications
- Add aiter fused mla
- Add aiter mha
- Add aiter customer all reduce
- Enable ep moe support
- Enable shared experts without ep
- Enable bias grouped topk
- Enable nextn

### TODO
- [ ] aiter branch clean up
- [ ] aiter rope embedding for DSV

### Pre-test
**Install aiter, refer to https://github.com/ROCm/aiter**
```
git clone --recursive https://github.com/BruceXcluding/aiter.git
python3 setup.py develop
```
**Before based on** https://github.com/sgl-project/sglang/pull/4075:
```
/sgl-workspace/sglang# RCCL_MSCCL_ENABLE=0 CK_MOE=1 python -m sglang.bench_one_batch --batch-size 64 --input 512 --output 32 --model deepseek-ai/DeepSeek-R1/ --tp 8 --trust-remote-code --quantization fp8
Benchmark ...
Prefill. latency: 2.35484 s, throughput:  13915.15 token/s
Decode.  latency: 0.03659 s, throughput:   1749.00 token/s
Decode.  latency: 0.03680 s, throughput:   1739.00 token/s
Decode.  latency: 0.03729 s, throughput:   1716.12 token/s
Decode.  latency: 0.03764 s, throughput:   1700.34 token/s
Decode.  latency: 0.03790 s, throughput:   1688.75 token/s
Decode.  median latency: 0.03876 s, median throughput:   1651.03 token/s
Total. latency:  3.549 s, throughput:   9810.84 token/s
```
**After: 
Prefill boost 19%
Decode boost 27%**
```
Benchmark ...
Prefill. latency: 1.91751 s, throughput:  17088.83 token/s
Decode.  latency: 0.02716 s, throughput:   2356.06 token/s
Decode.  latency: 0.02702 s, throughput:   2368.33 token/s
Decode.  latency: 0.02722 s, throughput:   2351.36 token/s
Decode.  latency: 0.02759 s, throughput:   2319.52 token/s
Decode.  latency: 0.02762 s, throughput:   2316.86 token/s
Decode.  median latency: 0.02829 s, median throughput:   2262.30 token/s
Total. latency:  2.792 s, throughput:  12471.83 token/s
```
**Accuracy**
```
python3 benchmark/gsm8k/bench_sglang.py --num-questions 2000 --parallel 2000 --num-shots 8
100%|███████████████████████████████████████████████████████| 1319/1319 [00:44<00:00, 29.64it/s]
Accuracy: 0.956
Invalid: 0.000
Latency: 49.351 s
Output throughput: 2688.740 token/s
```

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by zhyncs** [comment]
It looks so good!

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Great job! @BruceXcluding , could we add **"--aiter-backend"** so that we can use both aiter ops officially maintained by AMD and CUDA codes from SGLang team which is enabled in AMD platform ? 

I believe that we could have different ways to support AMD and AMD can use **aiter** to showcase the amzing performance as leaderboard in AMD platform with various AMD specific technology.

Otherwise we have to maintain a private SGLang branch for AMD platform, because we employ CUDA path first, but give them AMD friendly support.

**Comment by BruceXcluding** [comment]
> Great job! @BruceXcluding , could we add "--aiter-backend" so that we can use both aiter officially maintained by AMD and CUDA codes enabled and AMD platform ?
> 
> Otherwise we have to maintain a private SGLang branch for AMD platform, because we employ CUDA path first but support them in AMD friendly.

In the long run, we need an aiter backend, but now we control with flag "CK_MOE".

File: python/sglang/srt/_custom_ops.py [hunk 1]
```diff
@@ -77,42 +80,89 @@ def init_custom_ar(
             rank: int,
             full_nvlink: bool,
         ) -> int:
-            return sgl_kernel.allreduce.init_custom_ar(
-                meta, rank_data, handles, offsets, rank, full_nvlink
+            init_func = (
+                aiter_custom_ar.init_custom_ar
+                if get_bool_env_var("CK_MOE")
```
Comment (yiakwy-xpu-ml-framework-team) [inline python/sglang/srt/_custom_ops.py]: get_bool_env_var("CK_MOE") must be used together with **is_hip**, would it better to provide hip env function ?

```suggestion
                if get_hip_bool_env_var("CK_MOE")
```

File: python/sglang/srt/_custom_ops.py [hunk 2]
```diff
@@ -69,6 +69,9 @@ def register_graph_buffers(
     if is_hip():
```
Comment (yiakwy-xpu-ml-framework-team) [inline python/sglang/srt/_custom_ops.py]: suggested to use **_is_hip**

File: python/sglang/srt/_custom_ops.py (hunk 1 above)
Comment (yiakwy-xpu-ml-framework-team) [reply python/sglang/srt/_custom_ops.py]: The function can be local lambda by capturing **_is_hip** defined on the top of this module

File: python/sglang/srt/_custom_ops.py (hunk 2 above)
Comment (yiakwy-xpu-ml-framework-team) [reply python/sglang/srt/_custom_ops.py]: Note with use of **_is_hip** , ROCM codes should run either with atier backend or not. Perferable aiter with better performance.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 3]
```diff
@@ -176,6 +185,17 @@ def __init__(
             self.fp8_dtype = torch.float8_e4m3fn
             self.activation_scheme = quant_config.activation_scheme
 
+        if _is_hip and get_bool_env_var("CK_MOE"):
+            self.routed_scaling_factor = routed_scaling_factor
+            self.expert_mask = torch.zeros(
+                (self.num_experts + self.num_shared_experts + 1),
+                device="cuda",
+                dtype=torch.int,
+            )
```
Comment (yiakwy-xpu-ml-framework-team) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:215]: **torch.zeros** should happen in **__init__** codes, otherwise, it hurts performance significantly.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 4]
```diff
@@ -176,6 +185,17 @@ def __init__(
...
+            self.expert_mask[self.start_expert_id : self.end_expert_id + 1] = 1
+            self.expert_mask[self.num_experts : -1] = 1
```
Comment (yiakwy-xpu-ml-framework-team) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: The copy codes will be very expensive. Better to use inplace version of combination of **fill_**, **copy_**, **slice_** to initialize a tensor.

This can be automatically done if codes are wrapped inside torch.script.jit.

File: python/sglang/srt/models/deepseek_nextn.py [hunk 5]
```diff
@@ -222,6 +304,12 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                 # for mlp.experts[0].gate_gate_up_proj, which breaks load.
                 if ("mlp.experts." in name) and name not in params_dict:
                     continue
+                if (
+                    _is_hip
+                    and os.getenv("SGLANG_ROCM_AITER_BLOCK_MOE") == "1"
```
Comment (yiakwy-xpu-ml-framework-team) [inline python/sglang/srt/models/deepseek_nextn.py]: We must clarify when to use **SGLANG_ROCM_AITER_BLOCK_MOE** and **CK_MOE**

**Comment by yiakwy-xpu-ml-framework-team** [comment]
> > Great job! @BruceXcluding , could we add "--aiter-backend" so that we can use both aiter officially maintained by AMD and CUDA codes enabled and AMD platform ?
> > Otherwise we have to maintain a private SGLang branch for AMD platform, because we employ CUDA path first but support them in AMD friendly.
> 
> In the long run, we need an aiter backend, but now we control with flag "CK_MOE".

Thank you @BruceXcluding . I understand. Better to void use environment variable inside **forward**.

Let me give an explanation : get_env in posix platform is essentially implemented as char* parsing . It can be very slow for runtime codes.

I suggest you to add argument options in srt.arguments and use it later.

cc @HaiShaw

File: python/sglang/srt/models/deepseek_nextn.py (hunk 5 above)
Comment (BruceXcluding) [reply python/sglang/srt/models/deepseek_nextn.py]: @yiakwy-xpu-ml-framework-team  thanks, modified. forgot to remove the previous flag

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 4 above)
Comment (BruceXcluding) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: changed. appreciate it.

File: python/sglang/srt/layers/attention/triton_ops/decode_attention.py [hunk 6]
```diff
@@ -715,6 +719,19 @@ def decode_attention_fwd(
             sm_scale,
             logit_cap,
         )
+    elif _is_hip and get_bool_env_var("CK_MOE"):
+        # ROCM MLA
+        mla_decode_fwd(
```
Comment (jpvillam-amd) [inline python/sglang/srt/layers/attention/triton_ops/decode_attention.py:724]: Hey just a heads up this API doesn't match up with the aiter API looks like we are missing qo_indptr

Comment (jpvillam-amd) [reply python/sglang/srt/layers/attention/triton_ops/decode_attention.py:724]: and max_seqlen_q

**Comment by ShivamB25** [comment]
@BruceXcluding  why was this closed may i ask

**Comment by ShivamB25** [comment]
@jpvillam-amd is this still being worked on ?
